numpy/ma/tests/test_core.py

```diff
@@ -5017,6 +5017,28 @@ def test_astype():
     assert_(x_f2.mask.flags.f_contiguous)
 
 
+# This sort of test over types could be done with a fixture
```

Comment:
```python
dts = [np.dtype(dt_) for dt_ in '?bhilqpBHILQPefdgFD']

@pytest.parameterize('dt1', dts)
@pytest.parameterize('dt2', dts)
def test_astype_basic(d1, d2):
```

Reply:
Yes, that would work. What I am thinking of is having some standard fixtures for the various kinds in `testing`, but that needs discussion on the list. We tend to test to many special cases.

Reply:
The reason I didn't do it that way is I wanted the test to be self contained.

Reply:
> What I am thinking of is having some standard fixtures

Heck, even some standard lists of types would be useful. `np.sctypes` is useless because it's missing a bunch of types (#11923), `np.typeDict` and `np.sctypeDict` are awkward because they're full of useless aliases. Even using subclasses is awkward, because `inspect.isabstract(np.number)` doesn't return true.

Reply:
> I wanted the test to be self contained in the function.

I'd argue that the test is the contents of the inner for loop, not the whole block.

Reply:
Running a quick check, the output from testing `numpy/ma/tests/test_core.py` with default verbosity (dots) more than doubles in size from 4 to 9 lines.

Reply:
Is there a reason we're worried about the "size" of the testing output? More dots = isolated, granular tests with clearer associations on failures, seems ok to me? I'm always trying to minimize the number of assertions in a single unit test too.

It is a [contentious](https://softwareengineering.stackexchange.com/q/7823) software engineering topic it seems--arguments ranging from "just copy-pasting more if we do that" to purist arguments about single causes of failure for single tests.

Reply:
Also, printing to a file, takes .78 sec vs .09, so about 9x slower.

Reply:
I'm more worried about the time difference, which is significant. That is just for the one test.

Reply:
> Also, printing to a file, takes

I'd expect pytest to have more overhead in dispatching extra tests than printing `.` to the output does. How does the timing of the entire test file change with the `parametrize` decorators?

Reply:
1.87 vs 1.51 seconds. I'm not too concerned about the single test, but in using the parameterization in general. I think the time difference comes in setting up a separate test function for each iteration.

Reply:
So maybe the best way is with an internal loop in a try...except statement and generate a relevant error message on failure, although I see that pytest does show the parameter values when the test fails, which make having an error message less useful.

Reply:
Agreed with @eric-wieser here - `parametrize` is nice and makes (I think) the code clearer. Does the number of levels matter? I.e., is it faster to do
```
@parametrize('dt1, dt2', itertools.product(dt, dt))
```
(I know this works since I wrote code like the above - never thinking that one could just add two decorators...)

Reply:
Somewhat better, goes from  .78 sec to .57 seconds for the single tests. The time difference may not be due just to the parameterization, but to the fact that the inner loop is much smaller when parameterization is not used.

Reply:
Wait, the .57 is after the module is compiled, nested doesn't make any difference. So the choice is between .09 and .57 seconds. Also, fixtures have the same performance here, so I'm guessing what we are seeing is test setup overhead.

Due to popular demand, I'll push the parametrized version.